# Tasks

As part of The Emerging Technologies Module, we will be assigned four tasks regarding python to complete.

These are my solutions to the Tasks assesment. The author is Neil Byrne - G00343624
***

# Task 1: 
Write a Python function called sqrt2 that calculates and prints to the screen the square root of 2 to 100 decimal places. Your code shouldnot depend on any module from the standard library1or otherwise.  You shouldresearch the task first and include references and a description of your algorithm. 

Square Root: The square of a number x is, by definition, a number y such that y2=x.[4]

We can calculate the square root of a number by using Neton's method [1].
To find the square root $z$ of a number $x$, we can iterate using the following equation.

$$ z = z - z \frac{z^2 - x} {2z} $$


[1] A Tour of Go; Excercise: Loops and Functions; https://tour.golang.org/flowcontrol/8

[2] Newton's method; https://en.wikipedia.org/wiki/Newton%27s_method

[3] Newton's method coded ; https://stackoverflow.com/questions/46183020/square-root-without-pre-defined-function-in-python

[4] Square root of 100; https://socratic.org/questions/what-is-the-square-root-of-100

[5] F Strings in Python; https://realpython.com/python-f-strings/

In [13]:
def sqrt(n): # Initial guess of the square root of n

    x = int(n) 
    y = int(1)
      
    # Loop until we're happy with the accuracy
    a = 0.00001
    while(x - y > a): 
        x = (x + y) // 2
        y = n // x 
  
    return int(x)


  ##### Tests of the function
  Here we test the function with a known values.

Following this function we can calculate the square root as long as there is no remainder.
Python's floating points have a limit of what they will print out, calculating the square of a number 
to 100 places will not be possible this way.

Solution: We will make out number much bigger, to take out the desired result we just have to move the decimal place to the right point.

In [16]:
# Test the function on 100
sqrt(100)

10

$$ sqrt(100) = 10 $$
$$ 10^2 = 100 $$

In [38]:
x = 2 * 10 ** 200 # Making our number much bigger 
result =sqrt(x)
print(f'{result // 10**100}.{result % 10**100:100d}') # Using f strings we can maninpulate 
                                                      # our answer for the desired result [5]

1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727


In [27]:
# Check our answer 
from decimal import *
getcontext().prec = 101 # Amount of digits before and after the decimal
Decimal(2).sqrt()

Decimal('1.4142135623730950488016887242096980785696718753769480731766797379907324784621070388503875343276415727')

# Task 2:

The  Chi-squared  test  for  independence  is  a  statisticalhypothesis test like at-test.  
It is used to analyse whether two categorical variablesare independent.  The Wikipedia article gives the table below as an example [4],stating the Chi-squared value based on it is approximately 24.6. Usescipy.statsto verify this value and calculate the associatedpvalue.  You should include a shortnote with references justifying your analysis in a markdown cell.

Fortunately the scipy.stats class has a built in method to calculate the Chi-square[1]. From here we can then calculate the p value. 


[1] Scipy Docs https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chisquare.html

 #####  Chi-Square Test
This test is used when you are dealing with categoircal values rather
than numerical. It is used to determine whether or not there is a
correlelation between the associated values.

Here we test the function with some known values.

In [28]:
# Scipy.stats.chisquare is the import we will be using
from scipy.stats import chi2_contingency

In [29]:
# Imports 
import seaborn as sns
import pandas as pd
import numpy as np

In [30]:
# A dataframe saved container the Wikipedia example 
data = pd.read_csv('dataset.csv',header=1,sep=';',index_col=0)
df = pd.DataFrame(data)
df

A    B    C    D   Total
White Collar   90   60  104   95     349
Blue Collar    30   50   51   20     151
No Collar      30   40   45   35     150
Total         150  150  200  150     349

In [31]:
# For this sum we do not need the totals
# We will create a new table with no Totals

observedValues = np.array([[90, 60, 104, 95], # White collar
                  [30, 50,  51, 20], # Blue collar
                  [30, 40,  45, 35]]) # No collar

observedValues

array([[ 90,  60, 104,  95],
       [ 30,  50,  51,  20],
       [ 30,  40,  45,  35]])

SciPy has a built in function that will produce the Chi-Sqaure, degree of freedom and the expected array
https://docs.scipy.org/doc/scipy/reference/generated/scipy.stats.chi2_contingency.html
    

In [33]:
chi2, p, dof, expectedValues = chi2_contingency(observedValues)
# From the given population stats, this is the "expected results"
expectedValues

array([[ 80.53846154,  80.53846154, 107.38461538,  80.53846154],
       [ 34.84615385,  34.84615385,  46.46153846,  34.84615385],
       [ 34.61538462,  34.61538462,  46.15384615,  34.61538462]])

In [34]:
# In practice the expected and observed results are close but not exact
observedValues


array([[ 90,  60, 104,  95],
       [ 30,  50,  51,  20],
       [ 30,  40,  45,  35]])

In [35]:
# The calculated Chi-Square value
print(f"Chi-Square:{chi2:.10g}")

Chi-Square:24.57120286


This programming task was to verify that the example on the Wikipedia page [2] is accurate. 

[2] Chi-Sqaured Test https://en.wikipedia.org/wiki/Chi-squared_test

Wiki Answer ≈ 24.6

Scipy.stats Answer = 24.57120286

24.6 ≈ 24.57120286!

In [37]:
# The p value
print(f"p-value:{p:.5g}")

p-value:0.00040984


# Task 3:
The  standard  deviation  of  an  array  of  numbersxiscalculated usingnumpyasnp.sqrt(np.sum((x - np.mean(x))**2)/len(x)).However,  Microsoft  Excel  has  two  different  versions  of  the  standard  deviation calculation,STDEV.P and STDEV.S.  The STDEV.P function performs the above calculation but in the STDEV.S calculation the division is by len(x)-1 rather than len(x).  Research these Excel functions, writing a note in a Markdown cellabout  the  difference  between  them.   Then  use numpy to  perform  a  simulation demonstrating that the STDEV.S calculation is a better estimate for the standard deviation of a population when performed on a sample.  Note that part of this task is to figure out the terminology in the previous sentence.

##### The Standard Deviation of an array of numbers.

Within Microsoft Excel there are two different methods to calculate the standard deviation.

STDEV.P - $$ np.sqrt(np.sum((x - np.mean(x))**2)/len(x) $$
          $$ \sigma = \frac {\sqrt{\sum(x- \mu)^2}}{n} $$

Calculates the standard deviation of a population.

STDEV.S - $$ np.sqrt(np.sum((x - np.mean(x))**2)/len(x)-1 $$
          $$ s = \frac {\sqrt{\sum(x- \bar x)^2}}{n-1} $$
      
Calculates the standard deviation of a sample.

In [36]:
# Python code to demonstrate stdev() function 

# importing Statistics module 
import statistics 
import numpy as np

# creating a simple data - set 
sample = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20] 

# Standard Deviation on a population
print("Standard Deviation of sample is % s "
				% (statistics.stdev(sample))) 


# Standard Deviation on a sample
print("Standard Deviation of sample is % s ")
np.std(sample, ddof=1)


Standard Deviation of sample is 5.916079783099616 
Standard Deviation of sample is % s 


5.916079783099616

# Task 4:
Use scikit-learn to  apply k-means  clustering  to Fisher’s  famous  Iris  data  set.   You  will  easily  obtain  a  copy  of  the  data  set  on-line.  Explain in a Markdown cell how your code works and how accurate it might be, and then explain how your model could be used to make predictions of species of iris.